apps/web-client/src/lib/db/cr-sqlite/note.ts

```diff
@@ -50,6 +51,231 @@ export async function getAllInboundNotes(db: DB): Promise<InboundNoteListItem[]>
+
+	const updateQuery = `
+		UPDATE note
+		SET ${updateFields.join(", ")}
+		WHERE id = ?
+	`;
+
+	updateValues.push(id);
+
+	await db.exec(updateQuery, updateValues);
+}
+
+export async function commitNote(db: DB, id: number): Promise<void> {
+	const note = await getNoteById(db, id);
+	if (note?.committed) {
+		console.warn("Trying to commit a note that is already committed: this is a noop, but probably indicates a bug in the calling code.");
+		return;
+	}
+
+	const query = `
+		UPDATE note
+		SET committed = 1, committed_at = (strftime('%s', 'now') * 1000)
+		WHERE id = ?
+	`;
+
+	await db.exec(query, [id]);
+}
+
+export async function deleteNote(db: DB, id: number): Promise<void> {
+	const note = await getNoteById(db, id);
+	if (note?.committed) {
+		console.warn("Trying to delete a committed note: this is a noop, but probably indicates a bug in the calling code.");
+		return;
+	}
 	return db.exec("DELETE FROM note WHERE id = ?", [id]);
 }
+
+export async function addVolumesToNote(db: DB, noteId: number, volume: VolumeStock): Promise<void> {
+	const note = await getNoteById(db, noteId);
+	if (note?.committed) {
+		console.warn("Cannot add volumes to a committed note.");
+		return;
+	}
+
+	const { isbn, warehouseId, quantity } = volume;
+
+	const insertOrUpdateTxnQuery = `
+		INSERT INTO book_transaction (isbn, quantity, warehouse_id, note_id)
+		VALUES (?, ?, ?, ?)
+		ON CONFLICT(isbn, note_id, warehouse_id) DO UPDATE SET
+			quantity = book_transaction.quantity + excluded.quantity,
+			updated_at = (strftime('%s', 'now') * 1000)
+	`;
+
+	await db.exec(insertOrUpdateTxnQuery, [isbn, quantity, warehouseId, noteId]);
+}
+
+export async function getNoteEntries(db: DB, id: number): Promise<NoteEntriesItem[]> {
+	const query = `
+		SELECT
+			bt.isbn,
+			bt.quantity,
+			bt.warehouse_id AS warehouseId,
+			b.title,
+			b.price,
+			b.authors,
+			b.publisher
+		FROM book_transaction bt
+		LEFT JOIN book b ON bt.isbn = b.isbn
+		WHERE bt.note_id = ?
+		ORDER BY bt.updated_at DESC
+	`;
+
+	const result = await db.execO<{
+		isbn: string | null;
+		quantity: number;
+		warehouseId: number;
+		title?: string;
+		price?: number;
+		authors?: string;
+		publisher?: string;
+	}>(query, [id]);
+
+	return result;
+}
+
+export async function updateNoteTxn(
+	db: DB,
+	noteId: number,
+	curr: { isbn: string; warehouseId: number },
+	next: { warehouseId: number; quantity: number }
+): Promise<void> {
+	const note = await getNoteById(db, noteId);
+	if (note?.committed) {
+		console.warn("Cannot update transactions of a committed note.");
+		return;
+	}
+
+	const { isbn, warehouseId: currWarehouseId } = curr;
+	const { warehouseId: nextWarehouseId, quantity: nextQuantity } = next;
+
+	// Check if the transaction exists
+	const [existingTxn] = await db.execO<{ updated_at: number }>(
+		`SELECT updated_at FROM book_transaction WHERE note_id = ? AND isbn = ? AND warehouse_id = ?`,
+		[noteId, isbn, currWarehouseId]
+	);
+
+	if (!existingTxn) {
```

Comment:
Wouldn't it make more sense if this was an upsert? We might interpret this as "ISBN 97800000000 should have quantity 4 in this note", instead of "update the current value for this ISBN".

Reply:
This was the behaviour in the previous implementation, and I think it made sense when you think about the usage:
- use `addVolumes` when adding new/additional entries
- use `updateTxn` when you change a warehouse or manually insert the quantity (both using their respective fields)